libcontainer/cgroups/systemd/v1.go

```diff
@@ -332,6 +331,34 @@ func (m *legacyManager) GetStats() (*cgroups.Stats, error) {
 	return stats, nil
 }
 
+// hasDeviceRules checks cgroup device rules. Returns true if any rules
+// other than "allow all" exist. Used by freezeBeforeSet.
+func (m *legacyManager) hasDeviceRules() bool {
+	const allowAll = "a *:* rwm\n"
+
+	path := m.Path("devices")
+	// XXX: if Apply has not been called, m.paths are not initialized, so
+	// we have to figure the path. TODO: remove once fixed.
+	if path == "" {
+		var err error
+		path, err = getSubsystemPath(m.cgroups, "devices")
+		if err != nil {
+			// Should not happen. Assume we have rules.
+			return true
+		}
+	}
+	list, err := cgroups.ReadFile(path, "devices.list")
```

Comment:
So, this will incur some additional overhead in case devices.list contains more rules. Such case is quite common, e.g. on `runc update` (but not in k8s use case).

A way to optimize this is
```diff
@@ -347,7 +349,7 @@ func (m *legacyManager) hasDeviceRules() bool {
                        return true
                }
        }
-       list, err := cgroups.ReadFile(path, "devices.list")
+       fd, err := cgroups.OpenFile(path, "devices.list", os.O_RDONLY)
        if err != nil {
                if errors.Is(err, os.ErrNotExist) {
                        return false
@@ -356,7 +358,16 @@ func (m *legacyManager) hasDeviceRules() bool {
                logrus.Debug(err)
                return true
        }
-       return list != allowAll
+       defer fd.Close()
+
+       // Check contents matches allowAll, without reading more than needed.
+       buf := make([]byte, len(allowAll)+1) // 1 extra byte for io.EOF.
+       n, err := fd.Read(buf)
+       if n != len(allowAll) || err != io.EOF {
+               return true
+       }
+
+       return bytes.Equal(buf, []byte(allowAll))
 }
 
 // freezeBeforeSet answers whether there is a need to freeze the cgroup before
```

but it does not look very elegant. WDYT @mrunalp ?

Reply:
If we add some tests to `hasDeviceRules` where one can mock the file/input i think that would work. Would be interesting to run a benchmark to see how big the change in performance really is. :+1:

Reply:
I did a benchmark (with real `devices.list` files as I guess the kernel has some work to do to convert the internal rules representation to the text), and with a few files having "allow all" rules and a few files with 10-15 rules, the optimized code is 15% faster. I think this do not justify the decreased readability.

I also found 2 bugs in my proposed code (one is Read do not return io.EOF unless it's at EOF, the other is last check should be negated🤦🏻).

Based on all that, I am dropping this idea of optimizing the hell out of it.

For the reference, here's a benchmark results:
```
BenchmarkHasDeviceRulesOld-8   	   38041	     28395 ns/op
BenchmarkHasDeviceRulesNew-8   	   45121	     23767 ns/op
```

... and the benchmark itself in its entirety (you have to populate the paths yourself!):

```go
package x

import (
	"bytes"
	"errors"
	"os"
//	"fmt"
	"testing"
)

func hasDeviceRulesOld(file string) bool {
	const allowAll = "a *:* rwm\n"

	list, err := os.ReadFile(file)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return false
		}
		return true
	}
	return string(list) != allowAll
}

func hasDeviceRulesNew(file string) bool {
	const allowAll = "a *:* rwm\n"

	fd, err := os.Open(file)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return false
		}
		return true
	}
	defer fd.Close()

	// Check contents matches allowAll, without reading more than needed.
	buf := make([]byte, len(allowAll))
	n, _ := fd.Read(buf)
//	fmt.Printf("read n: %v, exp: %v, buf: %s\n", n, len(buf), buf)
	if n != len(buf) {
		return true
	}

	return !bytes.Equal(buf, []byte(allowAll))
}

func benchmarkHasDeviceRules(b *testing.B, fn func(string) bool) {
	tc := []struct {
		path string
		has  bool
	}{
		{"/sys/fs/cgroup/devices/s1/devices.list", true},
		{"/sys/fs/cgroup/devices/system.slice/systemd-logind.service/devices.list", true},
		{"/sys/fs/cgroup/devices/user.slice/devices.list", false},
	}

	for _, c := range tc {
		for i := 0; i < b.N; i++ {
			has := fn(c.path)
			if has != c.has {
				b.Fatalf("path %q: exp %v, got %v", c.path, c.has, has)
			}
		}
	}
}

func BenchmarkHasDeviceRulesOld(b *testing.B) {
	benchmarkHasDeviceRules(b, hasDeviceRulesOld)
}

func BenchmarkHasDeviceRulesNew(b *testing.B) {
	benchmarkHasDeviceRules(b, hasDeviceRulesNew)
}
```

Reply:
TL;DR: the complexity is not worth the savings.